File: src/rust/src/write.rs
//! Append-only writes.
//!
//! `fast_append` adds new data files and a new snapshot without rewriting
//! anything that already exists. Row-level deletes, overwrites and MERGE are out
//! of scope for 0.1.0, and are also absent from iceberg-rust itself.
//!
//! Incoming batches are realigned onto the table's own Arrow schema before being
//! written. A data frame exported from R carries no Iceberg field ids, and the
//! Parquet writer needs them, so matching by column name and casting where types
//! differ is what makes an ordinary data frame appendable.

use std::collections::HashMap;
use std::sync::Arc;

use arrow_array::RecordBatch;
use arrow_cast::CastOptions;
use arrow_schema::{Schema as ArrowSchema, SchemaRef};
use extendr_api::prelude::*;
use iceberg::spec::DataFileFormat;
use iceberg::transaction::{ApplyTransactionAction, Transaction};
use iceberg::writer::base_writer::data_file_writer::DataFileWriterBuilder;
use iceberg::writer::file_writer::ParquetWriterBuilder;
use iceberg::writer::file_writer::location_generator::{
    DefaultFileNameGenerator, DefaultLocationGenerator,
};
use iceberg::writer::file_writer::rolling_writer::RollingFileWriterBuilder;
use iceberg::writer::{IcebergWriter, IcebergWriterBuilder};
use parquet::basic::{Compression, GzipLevel, ZstdLevel};
use parquet::file::properties::WriterProperties;

use crate::arrow_bridge::import_reader;
use crate::errors::{RResult, ctx};
use crate::runtime::block_on;
use crate::table::RTable;

fn compression_from(name: &str) -> RResult<Compression> {
    Ok(match name.to_ascii_lowercase().as_str() {
        "zstd" => Compression::ZSTD(ZstdLevel::default()),
        "snappy" => Compression::SNAPPY,
        "gzip" => Compression::GZIP(GzipLevel::default()),
        "lz4" => Compression::LZ4_RAW,
        "uncompressed" | "none" => Compression::UNCOMPRESSED,
        other => {
            return Err(extendr_api::Error::Other(format!(
                "unknown compression {other:?}; expected one of \"zstd\", \
                 \"snappy\", \"gzip\", \"lz4\", \"uncompressed\""
            )));
        }
    })
}

/// Reshape a batch from R onto the table's Arrow schema.
///
/// Columns are matched by name, not by position, so column order in the data
/// frame does not matter. Types are cast when they differ, which is what lets an
/// R `Date` or `POSIXct` land in the right Iceberg type.
fn align_batch(batch: &RecordBatch, target: &SchemaRef) -> RResult<RecordBatch> {
    let incoming = batch.schema();

    // A column the table does not have is almost always a typo or a stale
    // data frame, and silently dropping it would lose data without a word.
    let mut unexpected: Vec<&str> = incoming
        .fields()
        .iter()
        .map(|f| f.name().as_str())
        .filter(|n| target.field_with_name(n).is_err())
        .collect();
    if !unexpected.is_empty() {
        unexpected.sort_unstable();
        let mut expected: Vec<&str> = target.fields().iter().map(|f| f.name().as_str()).collect();
        expected.sort_unstable();
        return Err(extendr_api::Error::Other(format!(
            "the data has {} column(s) that the table does not: {}.\n\
             Table columns are: {}.",
            unexpected.len(),
            unexpected.join(", "),
            expected.join(", ")
        )));
    }

    let mut columns = Vec::with_capacity(target.fields().len());
    for field in target.fields() {
        // Not "which the table requires": a table created from a data frame has
        // every field optional, so naming the Iceberg requiredness here would be
        // wrong in the common case. Every column has to be present regardless,
        // because guessing that an absent one meant NA turns a mistyped name
        // into a silent column of nulls.
        let idx = incoming.index_of(field.name()).map_err(|_| {
            extendr_api::Error::Other(format!(
                "the data is missing column {:?}, which the table has.\n\
                 Every column of the table must appear in the data; supply it as \
                 NA if there is no value for it.",
                field.name()
            ))
        })?;
        let column = batch.column(idx);

        let column = if column.data_type() == field.data_type() {
            column.clone()
        } else {
            // safe = false, deliberately. The default cast replaces a value it
            // cannot represent with a *null*: appending 3e9 to an `int` column,
            // or "abc" to any numeric one, would otherwise write NA over the
            // caller's data and commit it without a word. Refusing the append is
            // the only answer that does not silently lose data.
            arrow_cast::cast_with_options(
                column,
                field.data_type(),
                &CastOptions {
                    safe: false,
                    ..CastOptions::default()
                },
            )
            .map_err(|e| {
                extendr_api::Error::Other(format!(
                    "column {:?} is {} in the data but {} in the table, and the \
                     two cannot be reconciled: {e}",
                    field.name(),
                    column.data_type(),
                    field.data_type()
                ))
            })?
        };
        columns.push(column);
    }

    RecordBatch::try_new(target.clone(), columns)
        .map_err(|e| ctx("the data does not fit the table schema", e))
}

#[extendr]
fn rs_table_append(
    tbl: ExternalPtr<RTable>,
    stream_addr: &str,
    compression: &str,
    property_keys: Vec<String>,
    property_values: Vec<String>,
) -> RResult<ExternalPtr<RTable>> {
    if property_keys.len() != property_values.len() {
        return Err("internal error: snapshot property keys and values differ in length".into());
    }

    let table = &tbl.table;
    let metadata = table.metadata();
    let iceberg_schema = metadata.current_schema().clone();

    // Refused here, before a byte is written. icebergr writes through a plain
    // data-file writer and computes no partition tuples, so every file it
    // produces for a partitioned table carries an empty one -- which iceberg-rust
    // rejects at commit with "Partition value is not compatible with partition
    // type", by which point the Parquet files are already in the warehouse.
    // Nothing then references them, and this package has no maintenance operation
    // to remove them, so a failed append would leave litter behind on every
    // attempt as well as reporting the wrong thing.
    let spec = metadata.default_partition_spec();
    if !spec.fields().is_empty() {
        let by: Vec<String> = spec
            .fields()
            .iter()
            .map(|f| {
                let source = iceberg_schema
                    .field_by_id(f.source_id)
                    .map(|s| s.name.clone())
                    .unwrap_or_else(|| f.source_id.to_string());
                format!("{}({})", f.transform, source)
            })
            .collect();
        return Err(extendr_api::Error::Other(format!(
            "cannot append to {:?}: it is partitioned by {}, and icebergr 0.1.0 \
             writes only to unpartitioned tables.\n\
             An append would have to compute a partition value for every row, \
             which this version does not do. Write to this table with an engine \
             that supports partitioned writes; see icebergr_spec_support().",
            table.identifier().name(),
            by.join(", ")
        )));
    }

    // The Iceberg schema converted to Arrow carries the field-id metadata that
    // the Parquet writer needs.
    let target: SchemaRef = Arc::new(
        ArrowSchema::try_from(iceberg_schema.as_ref())
            .map_err(|e| ctx("could not convert the table schema to Arrow", e))?,
    );

    let mut reader = import_reader(stream_addr)?;
    let mut batches = Vec::new();
    let mut rows: usize = 0;
    for batch in reader.by_ref() {
        let batch = batch.map_err(|e| ctx("could not read the data from R", e))?;
        if batch.num_rows() == 0 {
            continue;
        }
        let batch = align_batch(&batch, &target)?;
        rows += batch.num_rows();
        batches.push(batch);
    }

    // Committing an empty snapshot would add a row to the table's history that
    // says nothing happened. Leave the table untouched instead; the R side
    // reports this.
    if rows == 0 {
        return Ok(ExternalPtr::new(RTable {
            catalog: tbl.catalog.clone(),
            table: table.clone(),
        }));
    }

    let writer_properties = WriterProperties::builder()
        .set_compression(compression_from(compression)?)
        .build();

    let location_generator = DefaultLocationGenerator::new(metadata)
        .map_err(|e| ctx("could not determine where to write data files", e))?;
    // A fresh uuid per append keeps file names from colliding across appends.
    let file_name_generator = DefaultFileNameGenerator::new(
        "icebergr".to_string(),
        Some(uuid::Uuid::new_v4().to_string()),
        DataFileFormat::Parquet,
    );

    let parquet_writer_builder =
        ParquetWriterBuilder::new(writer_properties, iceberg_schema.clone());
    let rolling_writer_builder = RollingFileWriterBuilder::new_with_default_file_size(
        parquet_writer_builder,
        table.file_io().clone(),
        location_generator,
        file_name_generator,
    );
    let data_file_writer_builder = DataFileWriterBuilder::new(rolling_writer_builder);

    let data_files = block_on(async {
        let mut writer = data_file_writer_builder.build(None).await?;
        for batch in batches {
            writer.write(batch).await?;
        }
        writer.close().await
    })
    .map_err(|e| ctx("could not write the data files", e))?;

    let snapshot_properties: HashMap<String, String> =
        property_keys.into_iter().zip(property_values).collect();

    let new_table = block_on(async {
        let tx = Transaction::new(table);
        let mut action = tx.fast_append().add_data_files(data_files);
        if !snapshot_properties.is_empty() {
            action = action.set_snapshot_properties(snapshot_properties);
        }
        let tx = action.apply(tx)?;
        tx.commit(tbl.catalog.as_ref()).await
    })
    .map_err(|e| ctx("could not commit the append", e))?;

    Ok(ExternalPtr::new(RTable {
        catalog: tbl.catalog.clone(),
        table: new_table,
    }))
}

extendr_module! {
    mod write;
    fn rs_table_append;
}

File: R/write.R
# Append-only writes, plus the minimum needed to bring a table into existence.

#' Create a namespace
#'
#' @param catalog An `icebergr_catalog` from [icebergr_catalog()].
#' @param namespace The namespace to create. Accepts `"db"` or `c("a", "b")`.
#'
#' @return `catalog`, invisibly.
#'
#' @examples
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' icebergr_create_namespace(catalog, "db")
#' icebergr_list_namespaces(catalog)
#' @export
icebergr_create_namespace <- function(catalog, namespace) {
  check_catalog(catalog)
  levels <- as_namespace(namespace)

  # An Iceberg catalog will not create "a.b" until "a" exists, so creating a
  # nested namespace in one call fails with "No such namespace". Walk the
  # parents and fill in the missing ones. The leaf is still created
  # unconditionally, so asking for a namespace that already exists is an error
  # rather than a silent no-op.
  for (i in seq_len(length(levels) - 1L)) {
    parent <- levels[seq_len(i)]
    if (!rs_namespace_exists(catalog$ptr, parent)) {
      rs_create_namespace(catalog$ptr, parent)
    }
  }

  rs_create_namespace(catalog$ptr, levels)
  invisible(catalog)
}

#' Create an Iceberg table
#'
#' The table's schema is taken from `data`, so an existing data frame is enough
#' to define one. Iceberg field ids are assigned automatically, since a data
#' frame has no concept of them.
#'
#' @param catalog An `icebergr_catalog` from [icebergr_catalog()].
#' @param table A table identifier, `"namespace.table"`. The namespace must
#'   already exist; see [icebergr_create_namespace()].
#' @param data A data frame whose columns define the schema. No rows are written;
#'   only the column names and types are used. An Arrow schema is also accepted.
#' @param location Where to store the table. `NULL` lets the catalog decide,
#'   which is almost always what you want.
#'
#' @return An `icebergr_table` handle for the new, empty table.
#'
#' @details
#' The table is created unpartitioned. Partitioned table creation, like partition
#' evolution, is out of scope for this version; see [icebergr_spec_support()].
#'
#' @examples
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' icebergr_create_namespace(catalog, "db")
#'
#' tbl <- icebergr_create_table(
#'   catalog, "db.events",
#'   data.frame(id = integer(), amount = double(), label = character())
#' )
#' icebergr_schema(tbl)
#' @export
icebergr_create_table <- function(catalog, table, data, location = NULL) {
  check_catalog(catalog)
  check_string(location, "location")
  ident <- parse_identifier(table)

  if (!is.data.frame(data) && !inherits(data, "nanoarrow_schema")) {
    abort("`data` must be a data frame, or an Arrow schema.")
  }

  # The schema object has to outlive the call: Rust only borrows it.
  holder <- export_schema(data)
  ptr <- rs_create_table(
    catalog$ptr, ident$namespace, ident$name, holder$addr,
    # A caller-supplied location may be a Windows path or a URI. Only the former
    # has separators to normalise, and Iceberg wants them forward.
    if (is.null(location)) NULL else as_iceberg_location(location)
  )
  new_icebergr_table(ptr, catalog)
}

#' Register an existing table with a catalog
#'
#' Points a catalog at a table that already exists on disk, by giving it the
#' table's metadata file. This is how a warehouse directory becomes visible to an
#' in-process `memory` catalog, which keeps no persistent registry of its own.
#'
#' @param catalog An `icebergr_catalog` from [icebergr_catalog()].
#' @param table A table identifier, `"namespace.table"`, to register it under.
#' @param metadata_location Path to the table's `metadata.json`.
#'
#' @return An `icebergr_table` handle.
#'
#' @examples
#' # Build a table, then re-attach it from a second catalog, as you would in a
#' # new session: a memory catalog keeps no registry between sessions.
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' icebergr_create_namespace(catalog, "db")
#' tbl <- icebergr_create_table(catalog, "db.events", data.frame(id = 1:3))
#' tbl <- icebergr_append(tbl, data.frame(id = 1:3))
#'
#' # Iceberg writes one metadata file per commit; the newest is the current
#' # state of the table.
#' files <- list.files(warehouse,
#'   pattern = "metadata\\.json$", recursive = TRUE,
#'   full.names = TRUE
#' )
#' newest <- files[order(file.mtime(files))][length(files)]
#'
#' reopened <- icebergr_catalog("memory", warehouse = warehouse)
#' icebergr_create_namespace(reopened, "db")
#' again <- icebergr_register_table(reopened, "db.events", newest)
#' icebergr_collect(again)
#' @export
icebergr_register_table <- function(catalog, table, metadata_location) {
  check_catalog(catalog)
  check_string(metadata_location, "metadata_location", allow_null = FALSE)
  ident <- parse_identifier(table)

  if (!file.exists(metadata_location)) {
    abort(paste0(
      "No metadata file at ", encodeString(metadata_location, quote = "\""), "."
    ))
  }

  ptr <- rs_register_table(
    catalog$ptr, ident$namespace, ident$name,
    as_iceberg_location(normalizePath(metadata_location, mustWork = TRUE))
  )
  new_icebergr_table(ptr, catalog)
}

#' Append rows to an Iceberg table
#'
#' Writes `data` as one or more new Parquet data files and commits a new
#' snapshot. Nothing already in the table is rewritten or removed.
#'
#' @param tbl An `icebergr_table` from [icebergr_table()].
#' @param data A data frame, or anything [nanoarrow::as_nanoarrow_array_stream()]
#'   accepts, such as an Arrow Table.
#' @param compression Parquet compression: `"zstd"` (the default), `"snappy"`,
#'   `"gzip"`, `"lz4"` or `"uncompressed"`.
#' @param properties Optional named character vector recorded in the new
#'   snapshot's summary, for provenance. Do not put credentials here: snapshot
#'   summaries are stored in table metadata and are readable by anyone who can
#'   read the table.
#'
#' @return An updated `icebergr_table` handle that sees the new snapshot. The
#'   handle passed in is unchanged, so reassign it: `tbl <- icebergr_append(tbl, x)`.
#'
#' @details
#' Columns are matched to the table by *name*, not position, so column order in
#' `data` does not matter. Types are cast where they differ from the table's, and
#' a column the table does not have is an error rather than being dropped
#' silently.
#'
#' Appending zero rows is a no-op: it warns, and returns the table unchanged
#' rather than committing an empty snapshot that records that nothing happened.
#'
#' The table must be unpartitioned. An append to a partitioned table would have
#' to compute a partition value for every row, which this version does not do, so
#' it is refused before any data is written rather than failing at the commit with
#' files already left in the warehouse. Partitioned tables can still be read; see
#' [icebergr_partitions()] and [icebergr_spec_support()].
#'
#' A table registered with [icebergr_register_table()] must also have been
#' registered from a metadata file named the way Iceberg names them,
#' `<version>-<uuid>.metadata.json`, because the next one is derived from that
#' name. Every engine writes conforming names; a renamed or hand-made file reads
#' fine and is refused here, again before anything is written.
#'
#' This is an append. Row-level deletes, overwrites and MERGE are not supported;
#' see [icebergr_spec_support()].
#'
#' @examples
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' icebergr_create_namespace(catalog, "db")
#'
#' events <- data.frame(id = 1:3, amount = c(1.5, 2.5, 3.5))
#' tbl <- icebergr_create_table(catalog, "db.events", events)
#' tbl <- icebergr_append(tbl, events)
#' icebergr_collect(tbl)
#' @export
icebergr_append <- function(tbl,
                            data,
                            compression = c("zstd", "snappy", "gzip", "lz4", "uncompressed"),
                            properties = NULL) {
  check_table(tbl)
  compression <- match.arg(compression)

  keys <- character()
  values <- character()
  if (!is.null(properties)) {
    # anyNA(names()) as well as anyNA(values): nzchar(NA) is TRUE, so an NA name
    # slipped past the emptiness check and failed inside Rust with a bare
    # "Must not be NA".
    if (!is.character(properties) || is.null(names(properties)) ||
      any(!nzchar(names(properties))) || anyNA(names(properties)) ||
      anyNA(properties)) {
      abort("`properties` must be a fully named character vector without NAs.")
    }
    keys <- names(properties)
    values <- unname(properties)
  }

  before <- rs_table_current_snapshot(tbl$ptr)

  # The stream must outlive the call: Rust drains it during the append.
  holder <- export_stream(data)
  ptr <- rs_table_append(
    tbl = tbl$ptr,
    stream_addr = holder$addr,
    compression = compression,
    property_keys = keys,
    property_values = values
  )

  # Reported from what Rust actually did rather than from nrow(data), which only
  # exists for a data frame: an Arrow stream carrying no batches used to be a
  # silent no-op.
  if (identical(rs_table_current_snapshot(ptr), before)) {
    warn("`data` has no rows; the table is unchanged and no snapshot was committed.")
  }

  new_icebergr_table(ptr, tbl$catalog)
}
